### PR TITLE
Add gatherer for Xen server

### PR DESCRIPTION
**Pull request by calancha**
## What does this PR change?
Add gatherer for Xen server

The following screenshots show:

- xen-orchestra web UI: the Ubuntu VM is power OFF
- console: the gatherer collects the current state of the VMs: Ubuntu VM is stopped.
- xen-orchestra web UI: the Ubuntu VM is ON.
- console: the gatherer collects again the VM states: Ubuntu VM now is running.

suse-on-ubuntu-off-Screenshot from 2021-05-31 08-33-51

console-suse-on-ubuntu-off-Screenshot from 2021-05-31 08-37-28

suse-on-ubuntu-on-Screenshot from 2021-05-31 08-48-04

console-suse-on-ubuntu-on-Screenshot from 2021-05-31 08-50-01

## Links
https://github.com/uyuni-project/uyuni/issues/3411

### CODE REVIEW

**Comment by calancha** [comment]
@agraul  Regarding the port I am unsure what to do.
Is it OK to use HTTP port 80 here?

File: virtual-host-gatherer/lib/gatherer/modules/Xen.py
Comment (mbologna) [inline virtual-host-gatherer/lib/gatherer/modules/Xen.py:114]: If this commented out code is no longer relevant, please remove it

**Comment by paususe** [comment]
> @agraul Regarding the port I am unsure what to do.
> Is it OK to use HTTP port 80 here?

I think you should make the port configurable, defaulting to HTTPS on 443.

**Comment by calancha** [comment]
> > @agraul Regarding the port I am unsure what to do.
> > Is it OK to use HTTP port 80 here?
> 
> I think you should make the port configurable, defaulting to HTTPS on 443.

Sure, it makes sense to use 443.  I was just worry that I don't know how to enable it in my dev enviroment to check if it works.

**Comment by paususe** [comment]
> > > @agraul Regarding the port I am unsure what to do.
> > > Is it OK to use HTTP port 80 here?
> > 
> > 
> > I think you should make the port configurable, defaulting to HTTPS on 443.
> 
> Sure, it makes sense to use 443. I was just worry that I don't know how to enable it in my dev enviroment to check if it works.

When I first looked into this, I used XCP-NG. You can use it with libvirt locally (not in PRV) if you enable nested virtualization on your machine. Alternatively, you can download the free version of XenServer but it's the same (just better SDK docs).

File: virtual-host-gatherer/lib/gatherer/modules/Xen.py
Comment (meaksh) [inline virtual-host-gatherer/lib/gatherer/modules/Xen.py:99]: One question: regardless of the port you specify, shouldn't we use `https` protocol here?

File: virtual-host-gatherer/lib/gatherer/modules/Xen.py
Comment (meaksh) [inline virtual-host-gatherer/lib/gatherer/modules/Xen.py:140]: Does "Xen AHV" really exist?

AFAIK `AHV` stands for `Acropolis Hypervisor` which is the Nutanix product.

File: virtual-host-gatherer/lib/gatherer/modules/Xen.py
Comment (meaksh) [inline virtual-host-gatherer/lib/gatherer/modules/Xen.py:118]: Same consideration about AHV here

File: virtual-host-gatherer/lib/gatherer/modules/Xen.py
Comment (meaksh) [inline virtual-host-gatherer/lib/gatherer/modules/Xen.py:122]: Just a comment here: I think we can get the actual value for `totalCpuThreads` from the XenAPI. Maybe you can take some inspiration from the Salt `xapi_virt` module, where something like this is also done: https://github.com/openSUSE/salt/blob/openSUSE-3002.2/salt/modules/xapi_virt.py#L319

File: virtual-host-gatherer/lib/gatherer/modules/Xen.py
Comment (paususe) [reply virtual-host-gatherer/lib/gatherer/modules/Xen.py:140]: I just checked on a Citrix XenServer Express installation I have here and this is what it reports:

You can see the base OS is a RHEL clone:

```
[root@xen ~]# ls -alh /etc/*release
-rw-r--r-- 1 root root  48 12 jun  2020 /etc/centos-release
-rw-r--r-- 1 root root 242 12 jun  2020 /etc/os-release
lrwxrwxrwx 1 root root  14 10 mar 23:13 /etc/redhat-release -> centos-release
lrwxrwxrwx 1 root root  14 10 mar 23:13 /etc/system-release -> centos-release
```

But it reports its name as "Citrix Hypervisor", and that's probably what we should report for the official Citrix Xen Hypervisor. If you want to go the extra mile, you can implement detecting whether it's Citrix or XCP-NG and reporting different 'os' but I don't think it's necessary:

```
[root@xen ~]# cat /etc/os-release
NAME="Citrix Hypervisor"
VERSION="8.2.0"
ID="xenenterprise"
ID_LIKE="centos rhel fedora"
VERSION_ID="8.2.0"
PRETTY_NAME="Citrix Hypervisor 8.2.0"
ANSI_COLOR="0;31"
HOME_URL="http://xenserver.org/"
BUG_REPORT_URL="https://bugs.xenserver.org/"
```

The `/etc/centos-release` is there but it's not CentOS:

```
[root@xen ~]# cat /etc/centos-release
Citrix Hypervisor release 8.2.0 (xenenterprise)
```

TL;DR: I would return `'os': 'citrixhypervisor'`, or `'os': 'xenhypervisor'` or alike.

File: virtual-host-gatherer/lib/gatherer/modules/Xen.py
Comment (paususe) [reply virtual-host-gatherer/lib/gatherer/modules/Xen.py:99]: I don't think so. If the user has configured an insecure transport (HTTP over port 80), that's what we need to use, otherwise it would fail. I don't think anyone would configure HTTPS on port 80 ¿?

File: virtual-host-gatherer/lib/gatherer/modules/Xen.py
Comment (calancha) [reply virtual-host-gatherer/lib/gatherer/modules/Xen.py:122]: @meaksh I cannot find in the response a field for this:
```
  "cpu_info": {
    "cpu_count": "4",
    "socket_count": "4",
    "vendor": "GenuineIntel",
    "speed": "2500.014",
    "modelname": "Intel Xeon E312xx (Sandy Bridge, IBRS update)",
    "family": "6",
    "model": "42",
    "stepping": "1",
    "flags": "fpu de tsc msr pae mce cx8 apic sep mca cmov pat clflush mmx fxsr sse sse2 ss syscall nx rdtscp lm constant_tsc rep_good nopl cpuid pni pclmulqdq ssse3 cx16 sse4_1 sse4_2 popcnt aes xsave avx hypervisor lahf_lm cpuid_fault ssbd ibrs ibpb stibp xsaveopt",
    "features": "9fba2223-0f8bfbff-00000001-2c100800",
    "features_pv": "1f89cbf5-96b82203-2991cbf5-00000003-00000001-00000000-00000000-00000000-00001000-8c000400-00000000-00000000-00000000-00000000-00000000",
    "features_hvm": "1f8bfbff-97ba2223-2d93fbff-00000403-00000001-00000002-00000004-00000000-00001000-8c000400-00000000-00000000-00000000-00000000-00000000",
    "features_hvm_host": "1f8bfbff-97ba2223-2d93fbff-00000403-00000001-00000002-00000004-00000000-00001000-8c000400-00000000-00000000-00000000-00000000-00000000",
    "features_pv_host": "1f89cbf5-96b82203-2991cbf5-00000003-00000001-00000000-00000000-00000000-00001000-8c000400-00000000-00000000-00000000-00000000-00000000"
  },
```

**Review by agraul — CHANGES_REQUESTED** [review]
Thanks for the PR @calancha !

I left a few comments and questions inline. Feel free to ignore my nitpicks if you want, they are not blocking the merge.

About the default port: I guess that depends, if XenAPI supports both http and https we could also support both and just add another parameter. The user would then provide the protocol in Uyuni. The default port could be 80 or 443, I'm leaning towards 443 as a "secure default" but either would work for me.

**Comment by calancha** [comment]
> > > I think you should make the port configurable, defaulting to HTTPS on 443.
> > 
> > 
> > Sure, it makes sense to use 443. I was just worry that I don't know how to enable it in my dev enviroment to check if it works.
> 
> When I first looked into this, I used XCP-NG. You can use it with libvirt locally (not in PRV) if you enable nested virtualization on your machine. Alternatively, you can download the free version of XenServer but it's the same (just better SDK docs).

It works as you expected.  The following screenshot is from a local xcp-ng:
- nested virtualization enabled
- the Ubuntu guest has trusted the self-signed certificate of the host.

after-trusted-self-signed-certificate-port-443-Screenshot from 2021-06-14 09-11-16

**Comment by paususe** [comment]
What are we missing here to merge?

**Review by paususe — APPROVED** [review]
LGTM

**Comment by agraul** [comment]
> What are we missing here to merge?

There are still two unresolved comments from my first review.

**Comment by calancha** [comment]
> There are still two unresolved comments from my first review.

I will get back to them next week; sorry, I had a very busy week.

File: virtual-host-gatherer/lib/gatherer/modules/Xen.py
Comment (calancha) [reply virtual-host-gatherer/lib/gatherer/modules/Xen.py:121]: Looking at the API doc:
https://xapi-project.github.io/xen-api/classes/host.html
```
(string → string) map cpu_configuration 
The CPU configuration on this host. May contain keys such as "nr_nodes", "sockets_per_node", "cores_per_socket", or "threads_per_core"
```
That 'may' suggests that this attribute is not guaranted to be there.  Exactly what I see in my test machine.
This code https://fossies.org/linux/salt/salt/modules/xapi_virt.py
returns `None` if the field is absent:
```
"cputhreads": _get_val(host_rec, ["cpu_configuration", "threads_per_core"]),
```
I would use the following:
```
host['cpu_configuration'].get('threads_per_core', 1)
```
to be consistent with the default we use (1) at `Kubernetes.py`

File: virtual-host-gatherer/lib/gatherer/modules/Xen.py
Comment (calancha) [reply virtual-host-gatherer/lib/gatherer/modules/Xen.py:156]: I do agree, but we are doing the same in the rest of the gatherers.  I prefer to keep aligned with teh code base..

**Review by agraul — APPROVED** [review]

**Comment by agraul** [comment]
@calancha one thing we forgot until today, could you please add a changelog entry to `virtual-host-gatherer/virtual-host-gatherer.changes` at the top of the file, e.g.
```
- Add Xen gatherer module

-------------------------------------------------------------------
Sat Dec 05 13:13:42 CET 2020 - mc@suse.de
```

**Review by meaksh — CHANGES_REQUESTED** [review]
One extra thing we need to do here before merging this PR is to fix the SPEC file in order to provide a new `virtual-host-gatherer-Xen` subpackage that includes this `Xen.py` module on it.

**Comment by calancha** [comment]
> @calancha one thing we forgot until today, could you please add a changelog entry to `virtual-host-gatherer/virtual-host-gatherer.changes` at the top of the file, e.g.
> 
> ```
> - Add Xen gatherer module
> 
> -------------------------------------------------------------------
> Sat Dec 05 13:13:42 CET 2020 - mc@suse.de
> ```

I have added the specification for the new module at `virtual-host-gatherer.spec`.
If I am not woring, the entry in the changes file should appear automatically.

**Comment by agraul** [comment]
> If I am not woring, the entry in the changes file should appear automatically.

I'm not aware of anything that would generate the changelog entry for the virtual-host-gatherer package. I'm probably missing something, do you know what's adding the entry @calancha ?

**Comment by calancha** [comment]
> > If I am not woring, the entry in the changes file should appear automatically.
> 
> I'm not aware of anything that would generate the changelog entry for the virtual-host-gatherer package. I'm probably missing something, do you know what's adding the entry @calancha ?

Probably you are right.
I was imaging that maybe, the IBS somehow creates the entry after bulding the package.
 Let me confirm it with Julio.

**Comment by calancha** [comment]
> > > If I am not woring, the entry in the changes file should appear automatically.
> > 
> > 
> > I'm not aware of anything that would generate the changelog entry for the virtual-host-gatherer package. I'm probably missing aging that maybe, the IBS somehow creates the entry after bulding the package.
> Let me confirm it with Julio.

**Comment by mbologna** [comment]
Reminder: check how we can install xenapi python library dependency via `Requires` in specfile.

File: virtual-host-gatherer/virtual-host-gatherer.spec
Comment (agraul) [inline virtual-host-gatherer/virtual-host-gatherer.spec:137]: Just one thing to remember, the `-Xen` subpackage will not be installable until we [package](https://en.opensuse.org/openSUSE:Packaging_Python#The_fast_and_automated_way) [XenAPI](https://pypi.org/project/XenAPI/).

File: virtual-host-gatherer/lib/gatherer/modules/Xen.py
Comment (agraul) [inline virtual-host-gatherer/lib/gatherer/modules/Xen.py:99]: Can we be sure that the library's output is stable or could the data parsing break when the API Version changes? Is there some kind of "we stay backwards compatible" policy we can rely on?

Hardcoding the version *might* be required if XenAPI reserves the right change its output.

**Review by agraul — COMMENTED** [review]
Thanks for adding the `Requires`! I'm not sure if the removal of the API version is a good thing, please see the question I left in-line.

File: virtual-host-gatherer/lib/gatherer/modules/Xen.py
Comment (calancha) [reply virtual-host-gatherer/lib/gatherer/modules/Xen.py:99]: > Can we be sure that the library's output is stable or could the data parsing break when the API Version changes?

"Nothing's ever for sure, John. That's the only sure thing I do know." Charles Herman

> Is there some kind of "we stay backwards compatible" policy we can rely on?
> Hardcoding the version _might_ be required if XenAPI reserves the right change its output.

Thanks, that's a good point.  Unfortunately, I won't have more time to work on this.
Feel free to take ownership on it, or discard it or whatever it makes more sense.

**Comment by meaksh** [comment]
As @agraul mentioned above, we do not have yet any `python-XenAPI` package at OBS/IBS. So the `virtual-host-gatherer-Xen` subpackage won't be installable in our SUSE/OpenSUSE distros.

Before merging this PR we would need to prepare this package, decide where to push it and if we really want to maintain it, and also fix the spec file so the `Xen` module is shipped by `-Xen` subpackage.

This would require we schedule some time to work on this, as Tino is not longer with us in the team.

/cc @admd

**Pull request closed without merging**
